lib/courier/scheduler.ex

```diff
@@ -0,0 +1,197 @@
+  the poller awakes to check for new messages to send simply set `interval` in the opts:
+
+      # opts/opts.exs
+      opts :my_app, MyApp.Mailer,
+        adapter: Courier.Adapter.Logger,
+        interval: 1_000 * 60 * 60  # awakes once an hour
+
+  The value for the interval is in milliseconds in accordance with the value that
+  [`:timer.sleep/1`](http://erlang.org/doc/man/timer.html#sleep-1)  expects.
+
+  ## Store
+
+  The store is where messages are kept until they are ready to be sent. The default
+  store is `Courier.Stores.Simple` and is just an Agent, storing messages in-memory.
+  This may not be ideal for your use-case. You can override the store in the opts:
+
+      # opts/opts.exs
+      opts :my_app, MyApp.ScheduledMailer,
+        adapter: Courier.Scheduler,
+        mailer: MyApp.DefaultMailer,
+        store: MyApp.OtherStore
+
+  The custom store will need respond to a certain API. Please see the documentation for `Courier.Store`
+  for details or look at the source code for `Courier.Stores.Agent`.
+  """
+
+  defmodule Worker do
+    use GenServer
+
+    def start_link(state) do
+      GenServer.start_link(__MODULE__, state, [])
+    end
+
+    def init(opts) do
+      {:ok, opts}
+    end
+
+    def handle_call({:deliver, message}, _from, [store: store, adapter: adapter, opts: opts] = state) do
+      store.delete(message)
+      adapter.deliver(message, opts)
+
+      {:noreply, state}
+    end
+  end
+
+  @doc false
+  def deliver(%Mail.Message{} = message, opts) do
+    store(opts).put({message, timestamp(opts)})
+  end
+
+  def children(opts) do
+    mailer = opts[:mailer]
+    adapter = opts[:adapter]
+    store = store(opts)
+    task_sup = Module.concat(mailer, TaskSupervisor)
+    opts = Keyword.put(opts, :task_sup, task_sup)
+
+    pool_name = Module.concat(mailer, Pool)
+    opts = Keyword.put(opts, :pool_name, pool_name)
+
+    [
+      Supervisor.Spec.supervisor(Task.Supervisor, [[name: opts[:task_sup]]]),
+      :poolboy.child_spec(opts[:pool_name], pool_opts(pool_name, opts), [store: store, adapter: adapter, opts: opts]),
+      Supervisor.Spec.worker(store, []),
+      Supervisor.Spec.worker(__MODULE__, [opts])
+    ]
+  end
+
+  defp pool_opts(name, opts) do
+    [
+      name: {:local, name},
+      worker_module: Worker,
+      size: opts[:pool_size] || 10,
+      max_overflow: 0
+    ]
+  end
+
+  @doc false
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @doc false
+  def init(opts) do
+    state = %{opts: opts, messages: %{}}
+    Process.send_after(self(), :poll, interval(opts))
+
+    {:ok, state}
+  end
+
+  @doc false
+  def handle_info(:poll, %{opts: opts} = state) do
+    state =
+      store(opts).all(past: true)
+      |> Enum.reduce(state, fn(message, state) ->
+        %{ref: ref} = Task.Supervisor.async_nolink(opts[:task_sup], fn ->
+          :poolboy.transaction(opts[:pool_name], fn(worker_pid) ->
```

Comment:
It won't be significant, but access on a keyword list has to walk the list which is a lot of overhead to be doing in a tight loop for every message when we know it's the same value every time.